casper/src/main/scala/coop/rchain/casper/util/comm/FairRoundRobinDispatcher.scala

```diff
@@ -0,0 +1,144 @@
+package coop.rchain.casper.util.comm
+
+import scala.collection.immutable.Queue
+
+import cats.Show
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+
+import FairRoundRobinDispatcher._
+import coop.rchain.shared.Log
+
+class FairRoundRobinDispatcher[F[_]: Sync: Log, S: Show, M](
+    filter: M => F[Dispatch],
+    handle: (S, M) => F[Unit],
+    queue: Ref[F, Queue[S]],
+    messages: Ref[F, Map[S, Queue[M]]],
+    retries: Ref[F, Map[S, Int]],
+    skipped: Ref[F, Int],
+    maxSourceQueueSize: Int,
+    giveUpAfterSkipped: Int,
+    dropSourceAfterRetries: Int
+) {
+  assert(maxSourceQueueSize > 0)
+  assert(giveUpAfterSkipped >= 0)
+  assert(dropSourceAfterRetries >= 0)
+
+  def dispatch(source: S, message: M): F[Unit] =
+    filter(message).flatMap {
+      case Handle =>
+        ensureSourceExists(source) >>
+          enqueueMessage(source, message) >>
+          handleMessages(source)
+      case Pass => handle(source, message)
+      case Drop => ().pure
+    }
+
+  private[comm] def ensureSourceExists(source: S): F[Unit] =
+    messages.get
+      .map(_.contains(source))
+      .ifM(
+        ().pure,
+        queue.update(_.enqueue(source)) *>
+          messages.update(_ + (source -> Queue.empty[M])) *>
+          retries.update(_ + (source  -> 0)) *>
+          Log[F].debug(s"Added ${source.show} to the dispatch queue")
+      )
+
+  private[comm] def enqueueMessage(source: S, message: M): F[Unit] =
+    messages.get
+      .map(_(source).size < maxSourceQueueSize)
+      .ifM(
+        messages.update(ps => ps.updated(source, ps(source).enqueue(message))) *>
+          retries.update(_.updated(source, 0)),
+        Log[F].debug(s"Dropped message from ${source.show}")
+      )
+
+  private[comm] def handleMessages(source: S): F[Unit] =
+    queue.get.flatMap { q =>
+      if (q.head == source) ().pure
+      else failure(q.head)
+    } >> handleNextMessage // because maybe gave up on failure
```

Comment:
Let's say we have three sources: A, B, C; giveUpAfterSkipped = 5. It's `A` turn but `B` has sent 4 messages cranking the `skipped` counter to 4. Then C sends a single message - it ends up in `failure(C)` and `giveUp(C)` but `C` is a bit innocent here. Am I reading it right? Is it possible?

Reply:
As you said, it's `A`s turn. If `C` sent the last message (to skip) then `q.head` is still `A`.